Implement eye dropper for paint editor

mewtaylor opened the pull request:
This implements #147.

**Expected Behavior:**
* eye dropper should hide when the mouse hovers off of the canvas
* eye dropper should deactivate when clicking outside the canvas
* items selected before activating the eye dropper should stay selected during eye dropping
* choosing the background does not choose transparent, it chooses white or gray

This also splits the `ColorPicker` into container/presentation.

@fsih @paulkaplan I'm down to review this together next week if we want, in person!

mewtaylor committed:
Implement eye dropper for paint editor

paulkaplan commented:
@mewtaylor there are a few lint failures here

paulkaplan commented:
Also, I'm getting an error when trying to load this in the GUI. Are you seeing the same thing?

```
Uncaught Error: Element ref was specified as a string (wrappedInstance) but no owner was set. You may have multiple copies of React loaded.
```

fsih commented on src/components/box/box.css:
```diff
@@ -0,0 +1,2 @@
+.box {
```
What does this do?

mewtaylor replied:
it was a copy directly from gui. I can remove it – I thought it might be good to mirror gui more directly, but that seems unnecessary with this

fsih replied:
Let's take it out if it's not doing anything

fsih commented on src/components/loupe/loupe.jsx:
```diff
@@ -0,0 +1,108 @@
+
+import styles from './loupe.css';
+
+const zoomScale = 3;
```
Nit: capital constants

fsih commented on src/containers/paint-editor.jsx:
```diff
@@ -38,14 +40,37 @@ class PaintEditor extends React.Component {
     componentDidMount () {
         document.addEventListener('keydown', this.props.onKeyPress);
     }
+    shouldComponentUpdate (nextProps, nextState) {
```
Did you need to add this to optimize for something in particular?

mewtaylor replied:
for optimizing the `setInterval` method to not re-render every time the color info is changed

fsih replied:
It's checking this.state.colorInfo !== nextState.colorInfo so wouldn't it still rerender?

mewtaylor replied:
This handles the case of a) hovering outside of the confines of the canvas, or b) if the loupe is paused in a particular place for a while. If we're not too concerned with performance in this case, then I can take it out though

fsih replied:
Can we check the colorInfo equality before we call setState instead of skipping the rendering in shouldComponentUpdate?

fsih commented on src/containers/paint-editor.jsx:
```diff
@@ -145,12 +170,58 @@ class PaintEditor extends React.Component {
+        }
+    }
+    startEyeDroppingLoop () {
+        const canvas = this.paintEditor.getWrappedInstance().canvas;
```
Instead of doing this, can you get the canvas directly by passing componentRef as a prop and having paint editor set the canvas? getWrappedInstance seems a bit hacky.

mewtaylor replied:
I think I'm just going to use `document.getElement` rather than either, due to another bug I'm running into (a bug with refs that @paulkaplan mentions above)

fsih replied:
I think it's discouraged to use document.getElement in react? Paul would know better. Maybe we can figure out the ref bug together without using document.getElement.

mewtaylor replied:
@fsih I changed it in a different way that prevents using `getElement` or `getWrappedInstance` – what do you think of the way it is now, with the canvas being a property rather than a state variable?

fsih commented on src/containers/paint-editor.jsx:
```diff
@@ -145,12 +170,58 @@ class PaintEditor extends React.Component {
+            this.setState({colorInfo: null});
+        }
+    }
+    startEyeDroppingLoop () {
```
paint-editor.jsx is getting kind of huge. I wonder if there's a good way to split code like this into other components or helper files before it gets unmanageable.

mewtaylor replied:
I agree that it's getting huge, and that it'd be good to split things up. The reason I had this be in here rather than elsewhere is mostly to handle the context of the `Loupe` component. This component needs to exist at the context of the paint editor – rather than at the context of the canvas, for example – because we defined the behavior of it as hiding/going away on hover/click outside of the canvas itself. 

Given this issue, it seemed necessary to me for the component to exist at the paint editor level rather than in some separate component. It probably can be moved out, though, if that's a dealbreaker.

If we're looking to make the paint editor component smaller, though, I think we could get some easier refactors out of breaking out `Edit Tools` and `Mode Tools` into their own components.

fsih replied:
That makes sense. Let's file a bug to break those out when we have time.

fsih commented on src/helper/eye-dropper.js:
```diff
@@ -0,0 +1,71 @@
+import paper from '@scratch/paper';
+
+const PAPER_WIDTH = 864;
+const PAPER_HEIGHT = 648;
```
Where do these numbers come from?

mewtaylor replied:
I had to look at the rendered canvas to see them – I can't find a reference anywhere in our code, but this appears to be something that paper is doing, and this is necessary to scale the loupe to make sure it's looking at the right part of the canvas when extracting a color.

fsih replied:
![image](https://user-images.githubusercontent.com/2855464/33907064-e8592038-df51-11e7-9d47-36315704d745.png)

The coordinates are really off for me. We'll probably have to figure out where these numbers are coming from

fsih commented on src/helper/eye-dropper.js:
```diff
@@ -0,0 +1,71 @@
+import paper from '@scratch/paper';
```
This file might make sense in helper/tools

fsih commented on src/reducers/eye-dropper.js:
```diff
@@ -0,0 +1,55 @@
+const ACTIVATE_EYE_DROPPER = 'scratch-paint/eye-dropper/ACTIVATE_COLOR_PICKER';
+const DEACTIVATE_EYE_DROPPER = 'scratch-paint/eye-dropper/DEACTIVATE_COLOR_PICKER';
+
+const initialState = {
```
Maybe this state should be part of the mode state?

mewtaylor replied:
I'd leave that up to you and @paulkaplan if you think it makes sense – moving this to `reducers/modes.js` would mean adding additional functionality to the mode state, since right now, across all modes, it simply changed the name of the mode. Including more stuff in there would be new, and seem to only apply to the eye dropper. If that's cool with you two, then I can do that for sure

fsih replied:
There's 2 places to look for whether the eyedropper is active now, this.active and mode = eyedropper, and I'm worried that they could get out of sync if we forget about one or the other. Alternatively, what if we don't change the mode when switching to the eyedropper? Are there issues with another tool mode being active while the eyedropper is running?

mewtaylor replied:
There are issues with another tool mode being active while the eye dropper is running – this was causing issues with mouse events, which was why we moved this to being a tool, rather than having its own mouse events. 

I'm fine with moving the data that's needed for the eye dropper into the mode itself as a resolution to the issue of it being in two places, but for a reason I can't seem to pin down, just activating the eye dropper as a tool without deactivating another first disables that tool when the eye dropper is then deactivated after it finishes.

Alternatively – I could move the "isActive" logic to just use the `mode`, but keep the eye dropper info logic in the separate reducer. That way, the activation logic is confined to one place, but the data is not cluttering up the `modes` reducer. Thoughts @fsih ?

fsih replied:
Re: the 2nd paragraph, after you deactivated the eye dropper, it might be that you have to call tool.activate() on the old tool.

(But don't worry about it since I think the other way we were discussing is probably better anyway)

fsih commented on src/helper/eye-dropper.js:
```diff
@@ -0,0 +1,71 @@
+        this.onMouseDown = this.handleMouseDown;
+        this.onMouseMove = this.handleMouseMove;
+
+        this.active = false;
```
Looks like this is unused?

fsih commented on src/helper/eye-dropper.js:
```diff
@@ -0,0 +1,71 @@
+
+        this.active = false;
+        this.canvas = canvas;
+        this.colorInfo = null;
```
This too

fsih commented on src/helper/eye-dropper.js:
```diff
@@ -0,0 +1,71 @@
+
+            const componentToString = c => {
+                const hex = c.toString(16);
+                return hex.length === 1 ? `0${hex}` : hex;
```
Can you add a comment to explain this bit?

fsih reviewed (changes requested):
Nice!

mewtaylor committed:
Changes from PR feedback

fixes lint and gui compatibility issues shown by @paulkaplan, and also does some cleanup to references and structure as requested by @fsih. Thanks to both of you for the reviews!

FYI – part of the canvas reference change made it so that the `PaintEditor` presentation component could become a classless declaration.

mewtaylor commented:
@paulkaplan should work now, thanks for seeing that – it was an issue with `react-intl` and its `withRef` argument in a library component somehow. And @fsih I made some of your requested changes and left comments for the others, let me know if you'd like to see any other changes based on what's here!

fsih commented on src/containers/paint-editor.jsx:
```diff
@@ -200,10 +278,14 @@ PaintEditor.propTypes = {
-    pasteOffset: state.scratchPaint.clipboard.pasteOffset
+    isEyeDropping: state.scratchPaint.color.eyeDropper.active,
+    pasteOffset: state.scratchPaint.clipboard.pasteOffset,
+    previousItems: state.scratchPaint.color.eyeDropper.previousItems,
```
Is this left over?

mewtaylor committed:
don’t change modes when activate eye dropper

It works well to keep track of the previously active tool in redux, and reactivate that tool when the eye dropper is finished. This takes `eye-dropper` out of the mode tool reducer too, which is nice.

mewtaylor committed:
use `pixelRatio` instead of scale constant

Turns out that the width/height paper sets on the canvas is tracked at scale to the width/height we set in the `pixelRatio` value, so use that to handle grabbing pixel color rather than a constant we set.

fsih commented on src/helper/tools/eye-dropper.js:
```diff
@@ -0,0 +1,68 @@
+import paper from '@scratch/paper';
+
+const LOUPE_RADIUS = 20;
```
Nice! This looks waaay better

mewtaylor committed:
remove `shouldComponentUpdate`

also do proper validation of `previousTool`

mewtaylor committed:
remove empty box css file

fsih commented:
I found a couple of bugs.

1. When you have something selected and you pick a color, an undo state doesn't get set/ the stage doesn't get updated. The color picker currently doesn't save undo state every time the color changes, but it should save it when it gets closed. So we should make sure FillColorIndicator's handleChangeFillColor gets called (to set the _hasChanged flag) and then componentWillReceiveProps gets called when it's closed (so that it calls onUpdateSvg)

2. When you zoom in, the loupe movement gets off:
![mouse offset](https://user-images.githubusercontent.com/2855464/34001069-00d1fb32-e0bd-11e7-9b65-afd434eb2425.gif)
(the zoom ratio is in paper.view.zoom)

mewtaylor commented:
@fsih hmm – `handleChangeFillColor` is what's called, as it's what is passed in as the `callback` value in redux, so `_hasChanged` should be getting changed... I'll look to see if there's a bug with that. And thanks for pointing out the zoom thing!

mewtaylor committed:
remove unused reference

mewtaylor committed:
accommodate zoom in eye dropper

mewtaylor committed:
don’t close popover on eye dropper click

and also update the hsv states from an eye dropper color change.

fsih commented on src/containers/color-picker.jsx:
```diff
@@ -35,25 +35,37 @@ class ColorPicker extends React.Component {
     }
-    componentWillReceiveProps () {
-        // Just a reminder, new props do not update the hsv state
+    componentWillReceiveProps (newProps) {
```
You should probably get @paulkaplan to check this color picker stuff

fsih reviewed (dismissed):
Awesome! LGTM

paulkaplan commented:
@mewtaylor looks great! A couple minor things.
1. With the stage picker, the cursor is hidden and the central square is filled in with the picked color. I think we should keep this consistent with that. If you want I can contribute this change directly to your branch, there is a minor complication because you can't use `cursor: none` on something with `pointer-events: none`, so it isn't as simple as a single line of CSS. It'll require adding a class to the canvas container (since that can receive pointer events).
2. The eyedropper icon is... extremely tiny. We should probably bump the size of that and the swatch to at least a touchable 40x40. That was my bad originally for making that transparent swatch so tiny. /cc @carljbowman 
3. I get a console error when the eyedropper starts up about `Warning: `NaN` is an invalid value for the `top` css style property.` That may or may not be related to...
4. The initial loupe state seems strange, when you click it an empty loupe is shown at the top left...
![initial-loupe-position](https://user-images.githubusercontent.com/654102/34046646-c42ed44c-e17b-11e7-9e61-b43cc547e388.gif)

paulkaplan commented on src/components/loupe/loupe.jsx:
```diff
@@ -0,0 +1,110 @@
+LoupeComponent.propTypes = {
+    colorInfo: PropTypes.shape({
+        color: PropTypes.shape({
+            r: PropTypes.number,
```
Looks like there is a disagreement about the color format. In the proptypes and line 55, it assumes `{r: r, g: g, b: b, a: a}`, but in line 60 here and in the eye-dropper-tool `getColorInfo` it assumes `[r, g, b, a]`. I don't have a preference either way, but since this component should ideally match the GUI version, I think we should use the object style.

mewtaylor committed:
updates from @paulkaplan’s feedback

1. Removes the cursour over loupe, and fills the loupe center’s color correctly
2. Fixes a console error and loupe initialization issue
3. Increases the size of the swatches

Thanks @paulkaplan!

mewtaylor committed:
add comment to colorInfo shape

mewtaylor committed:
small styling updates

make the swatches aligned in gui, and make sure the cursor shows when not eye dropping. Also moves eye dropper initialization into the helper. Thanks @paulkaplan!

paulkaplan commented:
Looks good to me! I did some testing on this integrated with the GUI, but let's make sure to emphasize this in our next internal test.
